File: strategy/generator/signal/flip/ce.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.smooth import SmoothATR
from core.models.source import SourceType
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class CeFlipSignal(Signal):
    type: SignalType = SignalType.CeFlip
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    period: Parameter = StaticParameter(22.0)
    smooth_atr: Parameter = StaticParameter(SmoothATR.SMMA)
    period_atr: Parameter = StaticParameter(22.0)
    factor: Parameter = StaticParameter(3.0)

File: ta_lib/strategies/signal/src/flip/ce.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use trail::f;
use trend::ce;

pub struct CeFlipSignal {
    source_type: SourceType,
    period: usize,
    smooth_atr: Smooth,
    period_atr: usize,
    factor: f32,
}

impl CeFlipSignal {
    pub fn new(
        source_type: SourceType,
        period: f32,
        smooth_atr: Smooth,
        period_atr: f32,
        factor: f32,
    ) -> Self {
        Self {
            source_type,
            period: period as usize,
            smooth_atr,
            period_atr: period_atr as usize,
            factor,
        }
    }
}

impl Signal for CeFlipSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period, self.period_atr)
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let (direction, _) = ce(
            &data.source(self.source_type),
            &data.atr(self.smooth_atr, self.period_atr),
            self.period,
            self.factor,
        );

        f!(direction)
    }
}
